### PR TITLE
docs: fixed faulty import in angular elements example

### PR DESCRIPTION
**Pull request by MrJean**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

The example A Popup Service containes a faulty import statement and use of function in `app.component.ts`, which results in a non working example.

Issue Number: N/A

## What is the new behavior?
The import statement including the function has been changed so this results in a working example.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by MrJean** [comment]
@gkalpak Let me look into that. I'm quite new to contributing to the project so still trying to find my way round the code base.

**Comment by gkalpak** [comment]
That's awesome, @MrJean! Feel free to ask here, if you have any questions or run into diffuculties or need more info/clarifications/guidance.

<sub>I know that figuring out the whole process as a first-time contributor can be overwhelming - we're here to help :smiley:</sub>

**Comment by MrJean** [comment]
@gkalpak currently having trouble running a single e2e test with the command `yarn example-e2e --filter=elements`. It keeps running the general e2e tests and so I currently was not able to verify my e2e tests.

Also, for https://github.com/angular/angular/pull/24840/commits/0e73c74690ef0f4ec7233645529af55aa8f0a2a5 I changed the type of `popupEl` to `any` because my linter kept complaining `message` wasn't a known property of that element. Which is in fact the `@input()` of the component on the `HTMLElement`. I don't feel this is a proper solution and I wonder if this could be done better?

File: aio/content/examples/elements/e2e/src/app.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    beforeAll(function () {
+        browser.get('');
+    });
+
+    it('should display popup component with content `Popup: Message`', function () {
+      const popupComponentButton = element.all(by.tagName('button')).get(0);
+      popupComponentButton.click();
+
+      const popupComponent =  element.all(by.tagName('popup-component')).get(0);
+      expect(popupComponent.innerHtml).toEqual('Popup: Message');
```
Comment (gkalpak) [inline aio/content/examples/elements/e2e/src/app.e2e-spec.ts]: Is `innerHtml` a thing?

File: aio/content/examples/elements/e2e/src/app.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+/* tslint:disable:quotemark */
+describe('Elements', function () {
+
+    beforeAll(function () {
+        browser.get('');
+    });
+
+    it('should display popup component with content `Popup: Message`', function () {
+      const popupComponentButton = element.all(by.tagName('button')).get(0);
+      popupComponentButton.click();
```
Comment (gkalpak) [inline aio/content/examples/elements/e2e/src/app.e2e-spec.ts]: Can you also verify that there is no `popup-component` before clicking the button (here and below)?

**Review by gkalpak — CHANGES_REQUESTED** [review]
Thx for working on this, @MrJean. I've left some comments and I will look into the e2e test issue.

File: aio/content/examples/elements/src/main.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// #docregion
```
[the remark this answers is not in the document]
Comment (MrJean) [reply aio/content/examples/elements/src/main.ts]: Removed this

File: aio/content/examples/elements/src/app/popup.service.ts [hunk 4]
```diff
@@ -40,7 +39,7 @@ export class PopupService {
   // This uses the new custom-element method to add the popup to the DOM.
   showAsElement(message: string) {
     // Create element
-    const popupEl = document.createElement('popup-element');
+    const popupEl: any = document.createElement('popup-element');
```
[the remark this answers is not in the document]
Comment (MrJean) [reply aio/content/examples/elements/src/app/popup.service.ts]: This is great! I did not know about `WithProperties`!

File: aio/content/examples/elements/e2e/src/app.e2e-spec.ts (hunk 1 above)
Comment (MrJean) [reply aio/content/examples/elements/e2e/src/app.e2e-spec.ts]: It seems it's not. I've replaced it with `getText()` https://www.protractortest.org/#/api?view=webdriver.WebElement.prototype.getText

**Review by gkalpak — CHANGES_REQUESTED** [review]
Almost there :smiley:
Can you, please, also take care of https://github.com/angular/angular/pull/24840#discussion_r202091680?

File: aio/content/examples/elements/src/index.html [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8">
+    <base href="/">
+    <title>Elements</title>
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <link rel="stylesheet" href="assets/sample.css">
```
[the remark this answers is not in the document]
Comment (MrJean) [reply aio/content/examples/elements/src/index.html]: @gkalpak I have removed this file

File: aio/tools/examples/shared/package.json [hunk 6]
```diff
@@ -22,6 +22,7 @@
     "@angular/common": "^6.0.0",
     "@angular/compiler": "^6.0.0",
     "@angular/core": "^6.0.0",
+    "@angular/elements": "^6.0.0",
```
[the remark this answers is not in the document]
Comment (MrJean) [reply aio/tools/examples/shared/package.json]: @gkalpak That is right, forgot to commit it

**Comment by MrJean** [comment]
@gkalpak Having some strange behaviour with the e2e tests. When checking wether the element is present, this seems to work by design (I hope so). But when I try to check the contents of the element it says the element with selector could not be found. I have added `xit` to the 2 tests so they don't break. Something tells me it has to do with the component/element processing the input and the browser not being ready yet?

File: aio/content/examples/elements/src/index.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/content/examples/elements/src/index.html]: It seems like you have removed the whole `index.html` file. I was only referring to this line:

```html
<link rel="stylesheet" href="assets/sample.css">
```

:grin:

**Comment by gkalpak** [comment]
I have pushed some changes to this PR:

- Did some minor refactoring in the `elements` example (mainly to resolve one of the testing issues - see below).
- Squashed your commits down to 2 (because you would need to do that anyway before merging).
- Added a couple of small fixes in `!fixup` commits. (These `!fixup` commits will be folded into yours while merging, so make sure you are OK with the contents :grin:)
- Addd a "work in progress" (WIP) commit with some test refactorings.

In case you're wondering there were two issues with the tests:

1. Since the popup would auto-hide after 2 secs (and since protractor tests are async), the popups would hide before being able to grab their text.
    _I fixed that by adding a "close" button and remove the auto-close-after-timeout behavior._
2. According to the spec, Custom Elements can only be used with ES2015 classes; i.e ES5 constructor functions won't work. (There are polyfills that allow using constructor functions.)
    By default, our examples are transpiled to ES5, which causes the popup element to break (the popup component works fine).
    _I still need to figure out the best way to fix this. We either need a special build for the `elements` example that transpiles to ES2015 or we need to include a polyfill. Off the top of my head, I lean towards the former for now and maybe later switch to polyfills and also explain the polyfill situation in the guide._

##
TODO:

- [ ] Fix issue 2 (see above).
- [ ] Add more tests (e.g. closing the popups, changing the message).

**Comment by MrJean** [comment]
@gkalpak Great input from your side! 

1. I'll add more tests, but for know the `popup-element` will need to be placed in `xit` since they fail due to wrong build run (es2015 build needed or polyfill). Can you agree on this part?

2. I'd go for the es2015 build, but not sure where to start on that one.

**Comment by MrJean** [comment]
@gkalpak While we're at it, I've added your explanation in https://github.com/angular/angular/pull/24840#issuecomment-405033821 to the docs.

**Comment by gkalpak** [comment]
Thx for the changes, @MrJean. I will temporarily take this over, since some of the fixes require changes in our docs generation infrastructure, which is not trivial :smiley:

I'll hand it back over to you as soon as I sort these out. Hang tight 😄 
(And thanks for all the work so far ❤️)

**Review by gkalpak — APPROVED** [review]
I have some suggestions for changes in 086771d03 ("docs(elements): explanation about getting the typings right"), but since it isn't related to the purpose of this PR (fix broken syntax and add tests to ensure it doesn't break again), how about this:

1. Remove that commit from this PR.
2. Get this PR merged.
3. Open a new PR adding the typings info.

**Comment by MrJean** [comment]
@gkalpak I've reverted the commit on the docs as asked.

**Comment by MrJean** [comment]
@gkalpak New PR opened with doc content changes #24925

**Comment by gkalpak** [comment]
Instead of having both the initial commit and the revert commit (which will be confusing for people looking at the commit history in the future), it would be better to just remove both commits altogether.

You can do this with `git rebase --interactive ...` (more info on [rewriting history][1]).
Or if these are the last two commits, you can `git reset --hard @~2`. 
(In both case you need to `git push --force-with-lease` once you are done.)

[1]: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

**Comment by MrJean** [comment]
@gkalpak I've misunderstood what had to be done. This should be OK now.

**Comment by gkalpak** [comment]
Note to caretaker:
The Travis failure is due to flakey Saucelabs and not related to this PR.

**Comment by MrJean** [comment]
Any idea why the CLA got unsigned?

**Comment by gkalpak** [comment]
Thx, @matsko, for taking care of the CLA.

##
@MrJean, [this commit][1] pretty much explains the issue. If a PR includes commits by people other than the PR submitter (e.g. your PR has my commits), googlebot can't verify that I'm OK with you submitting my commits in your PR (it doesn't know I added them), so we have to manually verify that it's OK.
(But the god news is it was able to verify both our CLAs :wink:)

BTW, great work on this one :100: It should be merged and live soon :tada:

[1]: https://github.com/angular/angular/pull/24840#issuecomment-405043690

**Comment by mathpaquette** [comment]
@MrJean @jbogarthyde will this example has a memory leak without removeEventListener to the 'closed' event?

**Comment by gkalpak** [comment]
@mathpaquette, your question is a little vague, but if I understand it correctly, there won't be a memory leak, becasue the element is removed from the DOM and (if there are no other references to it) it will be garbage collected (along with its event listeners).

**Pull request closed without merging**